File: lidar_processing/obstacle_extractor_node.py
import rclpy
from rclpy.node import Node

from sensor_msgs.msg import LaserScan
from robot_interfaces.msg import Obstacle, ObstacleArray

import numpy as np
import math


class ObstacleExtractor(Node):

    def __init__(self):
        super().__init__("obstacle_extractor")

        self.declare_parameter("scan_topic", "/scan")
        self.declare_parameter("obstacles_topic", "/obstacles")
        self.declare_parameter("target_frame", "base_footprint")

        scan_topic = self.get_parameter("scan_topic").value
        obstacles_topic = self.get_parameter("obstacles_topic").value

        self.create_subscription(LaserScan, scan_topic, self.scan_callback, 10)
        self.obstacle_pub = self.create_publisher(ObstacleArray, obstacles_topic, 10)

        # Matriz de transformación manual LIDAR -> base_footprint
        # Traslación según URDF: (0.3, 0, 0.5)
        self.tf_translation = np.array([0.3, 0.0, 0.5])

    def scan_callback(self, msg: LaserScan) -> None:
        ranges = np.array(msg.ranges)
        valid_idx = np.where(np.isfinite(ranges))[0]

        obstacles_msg = ObstacleArray()
        obstacles_msg.header.stamp = msg.header.stamp
        obstacles_msg.header.frame_id = "base_footprint"

        if len(valid_idx) == 0:
            self.obstacle_pub.publish(obstacles_msg)
            return

        clusters = self._cluster_indices(valid_idx)

        for i, cluster in enumerate(clusters):
            cluster = np.array(cluster, dtype=int)
            if len(cluster) == 0:
                continue

            cluster_ranges = ranges[cluster]
            angles = msg.angle_min + cluster * msg.angle_increment

            # Coordenadas en el frame del LIDAR
            xs = cluster_ranges * np.cos(angles)
            ys = cluster_ranges * np.sin(angles)

            # Centroide
            cx = float(np.mean(xs))
            cy = float(np.mean(ys))

            # Log centroide en LIDAR
            # self.get_logger().info(
            #    f"[Cluster {i}] Centroide en LIDAR: x={cx:.3f}, y={cy:.3f}"
            # )

            # Transformación manual a base_footprint
            bx = cx + self.tf_translation[0]
            by = cy + self.tf_translation[1]

            # self.get_logger().info(
            #     f"[Cluster {i}] Transformado a base_footprint: x={bx:.3f}, y={by:.3f}"
            # )

            # Crear Obstacle
            obs = Obstacle()
            obs.x = bx
            obs.y = by
            obs.distance = math.hypot(bx, by)
            obstacles_msg.obstacles.append(obs)

        self.obstacle_pub.publish(obstacles_msg)

    @staticmethod
    def _cluster_indices(indices):
        splits = np.where(np.diff(indices) > 1)[0] + 1
        return np.split(indices, splits)


def main():
    rclpy.init()
    node = ObstacleExtractor()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/obstacle_extractor_node.cpp
#include "lidar_processing/obstacle_extractor_node.hpp"

#include <cmath>
#include <numeric>
#include <string>
#include <vector>

namespace lidar_processing {

// ── Constructor
// ───────────────────────────────────────────────────────────────

ObstacleExtractorNode::ObstacleExtractorNode(
    std::shared_ptr<ITransformStrategy> strategy,
    const rclcpp::NodeOptions &options)
    : Node("obstacle_extractor", options), strategy_{std::move(strategy)} {
  declare_parameter("scan_topic", "/scan");
  declare_parameter("obstacles_topic", "/obstacles");

  const auto scan_topic = get_parameter("scan_topic").as_string();
  const auto obstacles_topic = get_parameter("obstacles_topic").as_string();

  scan_sub_ = create_subscription<sensor_msgs::msg::LaserScan>(
      scan_topic, 10,
      [this](const sensor_msgs::msg::LaserScan &msg) { scanCallback(msg); });

  obstacle_pub_ = create_publisher<robot_interfaces::msg::ObstacleArray>(
      obstacles_topic, 10);

  RCLCPP_INFO(get_logger(),
              "ObstacleExtractorNode ready | scan: %s → obstacles: %s",
              scan_topic.c_str(), obstacles_topic.c_str());
}

// ── Strategy setter
// ───────────────────────────────────────────────────────────

void ObstacleExtractorNode::setStrategy(
    std::shared_ptr<ITransformStrategy> strategy) {
  strategy_ = std::move(strategy);
}

// ── Scan callback
// ─────────────────────────────────────────────────────────────

void ObstacleExtractorNode::scanCallback(
    const sensor_msgs::msg::LaserScan &msg) {
  robot_interfaces::msg::ObstacleArray obstacles_msg;
  obstacles_msg.header.stamp = msg.header.stamp;
  obstacles_msg.header.frame_id = "base_footprint";

  if (!strategy_) {
    RCLCPP_WARN_ONCE(get_logger(), "No transform strategy set — scan ignored.");
    return;
  }

  // 1. Collect valid indices (finite range, above minimum threshold)
  std::vector<int> valid_indices;
  valid_indices.reserve(msg.ranges.size());

  for (int i = 0; i < static_cast<int>(msg.ranges.size()); ++i) {
    const float r = msg.ranges[i];
    if (std::isfinite(r) && r > 0.1f) {
      valid_indices.push_back(i);
    }
  }

  if (valid_indices.empty()) {
    obstacle_pub_->publish(obstacles_msg);
    return;
  }

  // 2. Cluster contiguous indices
  const auto clusters = clusterIndices(valid_indices);

  for (const auto &cluster : clusters) {
    if (cluster.empty()) {
      continue;
    }

    // 3. Compute centroid in LIDAR frame
    double sum_x = 0.0;
    double sum_y = 0.0;

    for (const int idx : cluster) {
      const double angle = msg.angle_min + idx * msg.angle_increment;
      sum_x += msg.ranges[idx] * std::cos(angle);
      sum_y += msg.ranges[idx] * std::sin(angle);
    }

    const double cx = sum_x / static_cast<double>(cluster.size());
    const double cy = sum_y / static_cast<double>(cluster.size());

    // 4. Delegate transformation to the injected strategy
    const auto point =
        strategy_->transform(cx, cy, msg.header.stamp, msg.header.frame_id);

    if (!point.has_value()) {
      continue; // Transform unavailable this cycle; skip cluster silently
    }

    // 5. Build and append the obstacle
    robot_interfaces::msg::Obstacle obs;
    obs.x = point->x;
    obs.y = point->y;
    obs.distance = std::hypot(point->x, point->y);
    obstacles_msg.obstacles.push_back(obs);
  }

  obstacle_pub_->publish(obstacles_msg);
}

// ── Helpers
// ───────────────────────────────────────────────────────────────────

std::vector<std::vector<int>>
ObstacleExtractorNode::clusterIndices(const std::vector<int> &valid_indices) {
  std::vector<std::vector<int>> clusters;

  if (valid_indices.empty()) {
    return clusters;
  }

  clusters.push_back({valid_indices.front()});

  for (std::size_t i = 1; i < valid_indices.size(); ++i) {
    if (valid_indices[i] - valid_indices[i - 1] > 1) {
      clusters.push_back({}); // Gap detected → start new cluster
    }
    clusters.back().push_back(valid_indices[i]);
  }

  return clusters;
}

} // namespace lidar_processing
